### DISCUSSION TITLE
Probleme mit Sysvar/Program Beschreibungen

### DISCUSSION THREAD

**Discussion by SukramJ** [SukramJ/aiohomematic#1906] [General]
Sofern die Beschreibungen von Systemvariablen und Programmen unerlaubte Zeichen (" und \ \) enthalten, können diese nicht mehr geladen werden.

Mit den folgenden beiden Skripten können die bei der CCU abgefragten Systemvariablen untersucht werden.

Bitte die Beschreibungen auf nicht erlaubte Zeichen kontrollieren.

EDIT: Es sind nur die Zeichen " und \ \ problematisch!!

Ein Testskript findet ihr [hier](https://github.com/danielperna84/hahomematic/discussions/1906#discussioncomment-11536032).

**Comment by eikowagenknecht** [comment]
Ausführung per "Programme und Verknüpfungen" > "Skript testen".

Alles zusammen in einer Liste:

<details>

<summary>Script</summary>


```tcl
var svList = dom.GetObject(ID_SYSTEM_VARIABLES);
var prgList = dom.GetObject(ID_PROGRAMS);
string id;

Write("VARS:")
WriteLine('');
foreach(id, svList.EnumIDs())
{
    var sv = dom.GetObject(id);
    string description;
    if (sv)
    {
      var desc = sv.DPInfo();
      if (desc) {
        description = desc;
      } else {
        description = "";
      }

      Write("name: " # sv.Name() # "");
      WriteLine('');
      Write("description: " # description # "");
      WriteLine('');
    }
}

WriteLine('');
WriteLine('');
Write('PROGRAMS');
WriteLine('');

foreach(id, prgList.EnumIDs())
{
    var prg = dom.GetObject(id);
    string description;

    if (prg)
    {
      var desc = prg.PrgInfo();
      if (desc) {
        description = desc;
      } else {
        description = "";
      }

      Write("name: " # prg.Name() # "");
      WriteLine('');
      Write("description: " # description # "");
      WriteLine('');
    }
}
```
</details>


Um dann rauszufinden, wo falsche Werte drin sind, z.B.

Alles in Editor packen, suchen nach Regex: [<>'"\&\$\[\]\{\}\\]

@SukramJ 

Bei mir sind die folgenden mit nicht erlaubten Zeichen:
- name: CcuImReboot, description: Umgewidmete Variable "Anwesenheit"
  - kann ich selbst ändern, check
- name: WatchDog: high-temp, description: high system temperature (> 80 deg) detected
  - Systemerstellt, wie soll man das verhindern? Generiert hier: https://github.com/jens-maus/RaspberryMatic/blob/bbf9ad79df582f6fb03f6a2aaf182a5c14e1578a/buildroot-external/overlay/base-raspmatic/etc/monitrc#L387
- name: WatchDog: lighttpd-expired-sslcert, description: HTTPS certificate expiring in < 30 days
  - dito, generiert hier: https://github.com/jens-maus/RaspberryMatic/blob/bbf9ad79df582f6fb03f6a2aaf182a5c14e1578a/buildroot-external/overlay/base-raspmatic_oci/etc/monitrc#L76

Was ist der Hintergrund dieser neuerdings nicht mehr erlaubten Zeichen? Mir hat es heute Morgen meine Haussteuerung lahmgelegt bis ich die beta Version aus HACS installiert habe.

**Reply by SukramJ** [reply]
> Was ist der Hintergrund dieser neuerdings nicht mehr erlaubten Zeichen?

Die Beschreibungen der Systemvariablen/Programme werden in den Attributen angezeigt.

Interessant ist, das solche Beschreibungen (`> 80 deg`, `HTTPS certificate expiring in < 30 days`) über die UI (zumindest RM)  nicht angelegt werden dürfen.

**Comment by Baxxy13** [comment]
Da es ja primär darum geht "verbotene Zeichen" (in den Beschreibungen) aufzuspüren habe ich das Script von @eikowagenknecht etwas "umgebaut". 
Kein externer Regexer mehr nötig und es listet jetzt nur die problematischen SysVars/Programme auf und zeigt auch gleich die "verbotenen Zeichen".
(im "Script testen" auf der RM / CCU ausführen)

<details>

<summary>Script</summary>

```
string id;
string sDeadList = ^"	\^;
string sDeadSign;
integer iCount_SV;
integer iCount_PRG;
integer iCount;

Write("\tSYSTEMVARIABLEN:\n");
foreach(id, dom.GetObject(ID_SYSTEM_VARIABLES).EnumIDs()) {
  object sv = dom.GetObject(id);
  iCount = 0;
  if (sv) {
    string sSV_Desc = sv.DPInfo();
    foreach(sDeadSign, sDeadList) {
      if (sSV_Desc.Contains(sDeadSign)) {
        iCount = iCount + 1;
        iCount_SV = iCount_SV + 1;
        if (iCount == 1) {
          Write("\t---------------\n");
          WriteLine("\tName: " # sv.Name() #"\n\tdescription: "#sv.DPInfo());
        }
        WriteLine("WARNUNG: "#sDeadSign#" gefunden!!!");
      }
    }
  }
}
WriteLine("\n\tPROGRAMME:");
foreach(id, dom.GetObject(ID_PROGRAMS).EnumIDs()) {
  object prg = dom.GetObject(id);
  iCount = 0;
  if (prg) {
    string sPrg_Desc = prg.PrgInfo();
    foreach(sDeadSign, sDeadList) {
      if (sPrg_Desc.Contains(sDeadSign)) {
        iCount = iCount + 1;
        iCount_PRG = iCount_PRG + 1;
        if (iCount == 1) {
          Write("\t---------------\n");
          WriteLine("\tName: " # prg.Name() #"\n\tdescription: "#prg.PrgInfo());
        }
        WriteLine("WARNUNG: "#sDeadSign#" gefunden!!!");
      }
    }
  }
}
WriteLine("\nAnzahl gefundener problematischer Zeichen: "# iCount_SV + iCount_PRG);
WriteLine("in SysVar-Beschreibungen: "# iCount_SV);
WriteLine("in Programm-Beschreibungen: "# iCount_PRG); 

```

</details>

EDIT SukramJ: Ich hab das Skript auf die wirklich problematischen Zeichen reduziert.

**Reply by SukramJ** [reply]
Die `sDeadList` resultiert ja aus der UI Warnung, das diese Zeichen nicht verwendet werden dürfen.
Das Problem besteht aber eigentlich darin, das das aus dem [Script](https://github.com/danielperna84/hahomematic/tree/devel/hahomematic/rega_scripts#:~:text=2%20years%20ago-,get_system_variable_descriptions.fn,-Add%20description%20to) resultierende JSON nicht mehr gültig ist.
Bleibt dann im Idealfall nicht nur `"` und  `\` als ungültige Zeichen übrig?

Edit `\` ergänzt.

**Reply by Baxxy13** [reply]
Jap, `"` und `\` sind ungültig. Bin das gerade mal mit nem JSON-Validator durchgegangen.

Beide dürften äußerst selten in den Beschreibungen auftauchen. Das muss man schon "an der WebUI vorbei" per Script/SDV etc. sowas reinschreiben.

**Reply by SukramJ** [reply]
Ich hab mir mal erlaubt das Skript auf die problematischen Zeichen anzupassen.

**Reply by dennis-heissmann** [reply]
Hi, ich hab mit eurem Skript in meiner CCU3 geschaut ob dort Programme oder Variablen mit den entsprechenden Sonderzeichen drin sind und hab dabei zwei System-Variablen gefunden die Probleme machen. Allerdings werden die nicht unter den Systemvariablen selbst angezeigt, hat einer von euch evtl. das selbe Problem und bereits eine Lösung dafür?

 	SYSTEMVARIABLEN:

	---------------
	Name: ${sysVarAlarmMessages}
	description: ${sysVarAlarmMessagesCount}
WARNUNG: $ gefunden!!!
WARNUNG: { gefunden!!!
WARNUNG: } gefunden!!!
	---------------
	Name: ${sysVarServiceMessages}
	description: ${sysVarServiceMessagesCount}
WARNUNG: $ gefunden!!!
WARNUNG: { gefunden!!!
WARNUNG: } gefunden!!!

	PROGRAMME:

Anzahl gefundener problematischer Zeichen: 6
in SysVar-Beschreibungen: 6
in Programm-Beschreibungen: 0

**Reply by SukramJ** [reply]
Das Script von @Baxxy13? Dann sollten in der aktuellen Version die Fehler nicht mehr auftauchen.

**Reply by SukramJ** [reply]
Und du must dir die internen Programme/ Systemvariablen anzeigen lassen.

**Reply by Baxxy13** [reply]
Das ist unter dem Begriff "Klarnamenbug" bekannt. 
In RaspberryMatic lange ausgemerzt (siehe: https://github.com/jens-maus/RaspberryMatic/issues/199 ), in der CCU scheinbar nicht.

Diese Zeichen sind ja im aktuellen Script nicht mehr drin, daher grundsätzlich kein Problem. 
Auf der CCU kommt man an diese SysVars nicht ohne Scripting dran.

**Reply by dennis-heissmann** [reply]
> Und du must dir die internen Programme/ Systemvariablen anzeigen lassen.

Hab ich gemacht die werden aber dennoch nicht angezeigt wie @Baxxy13  schon geschrieben hatte.

> Das ist unter dem Begriff "Klarnamenbug" bekannt. In RaspberryMatic lange ausgemerzt (siehe: [jens-maus/RaspberryMatic#199](https://github.com/jens-maus/RaspberryMatic/issues/199) ), in der CCU scheinbar nicht.
>
> Diese Zeichen sind ja im aktuellen Script nicht mehr drin, daher grundsätzlich kein Problem. Auf der CCU kommt man an diese SysVars nicht ohne Scripting dran.

Danke, dass hatte ich auch gesehen. Allerdings nutze ich nicht Raspberry Matic sondern die richtige CCU3 von eq3.


Ich erhalte massen an Fehler im Log wenn ich von 1.72 auf 1.73 gehe. 

`2024-12-11 22:18:14.817 ERROR (MainThread) [homeassistant.components.update] Platform homematicip_local does not generate unique IDs. ID homematicip_local_000a18a9a81fe2_update already exists - ignoring update.heizung_og_kuche_update`
Solch eine Meldung erhalte ich bei allen entities.

Im UI zeigt es nur an, dass es einen "Einrichtungsfehler" gab
![image](https://github.com/user-attachments/assets/ce46b335-4f8d-47b1-8894-981c3c9b5ffe)

**Reply by SukramJ** [reply]
Macht aber auch Sinn auf die aktuelle Version zu wechseln.

**Reply by dennis-heissmann** [reply]
> Macht aber auch Sinn auf die aktuelle Version zu wechseln.

Ja klar, hab's jetzt nochmal probiert - jetzt hat es geklappt. Allerdings bekomme ich jetzt den klassischen Fehler

`ERROR (MainThread) [hahomematic.client.json_rpc] GET_PROGRAM_DESCRIPTIONS failed: Unable to decode json: unexpected control character in string: line 1 column 32 (char 31). Descriptions must not contain any of the following characters: <, >, ', ", &, $, [, ], {, }, \`

Das kann ich mir nicht ganz erkenne, das sind meine Programme in der CCU3 inkl. der internen:
![2](https://github.com/user-attachments/assets/8a20817b-2ac4-476c-8531-e9215d112c7b)

**Reply by SukramJ** [reply]
In der nächsten Version wird [dieses Script](https://github.com/danielperna84/hahomematic/blob/devel/hahomematic/rega_scripts/get_program_descriptions.fn) verwendet.
Ungültige Beschreibungen werden dann durch ein "Ignored ..." ersetzt. Vielleicht vereinfacht das die Suche.
Bitte das Ergebnis mal hier posten.

Ansonsten würde erst mal alle Beschreibungen komplett leeren, und dann schauen ob der Fehler noch kommt.

**Comment by spali** [comment]
Hatte das Problem mit eine Tab Zeichen das vermutlich durch Copy&Paste in eine Beschreibung gelangt ist.
Das Script von @Baxxy13  hat mir das nicht angegeben, obwohl ich überprüft habe das das Tab Zeichen in `sDeadList` auch in den Test-Script Editor richtig kopiert wurde... 
Gefunden habe ich die Problematische Beschreibung am Ende mit https://github.com/danielperna84/hahomematic/blob/devel/hahomematic/rega_scripts/get_system_variable_descriptions.fn und dann einem Online JSON-Validator.

**Reply by Baxxy13** [reply]
Das Script so wie es ist kann keine Tabs erkennen, die dienen in der Liste als Separator für die foreach-Schleife. 
Ob es generell möglich ist auf diese Weise Tabs zu erkennen weiß ich aktuell nicht, guck's mir aber bei Gelegenheit an.

**Reply by spali** [reply]
Habe es gerade getestet. Unschön ist, RM motzt nicht wenn ich per copy&paste ein Tab rein kopiere und speichere 😟 
Ev. würde es generell Sinn machen die Fehlermeldung im Log bei HomeAssistant entsprechend detaillierter zu machen... so dass man den Fehler auch ohne Skripts findet. Weil dies kann nun nachträglich jederzeit wieder passieren bei Copy&Paste wo aus versehen ein Tab mit kommt. Oder man ignoriert diese einfach gleich vorweg (raus filtern sollte zu mindestens in der Beschreibung keine Probleme machen).

**Reply by Baxxy13** [reply]
Also ich jetzt noch ein bisschen getestet aber keine Möglichkeit gefunden einen `Tab` direkt  als Fehlerzeichen auszugeben.
Aber ich könnte `Tabs` im String replacen, z.B. mit 2 oder 3 "normalen" Leerzeichen.

**Reply by SukramJ** [reply]
Das hört sich nicht schlecht an aber, ich denke das ein Leerzeichen reicht.

**Reply by SukramJ** [reply]
Das sieht dann in etwa so #1910 aus.

**Comment by jens-maus** [comment]
Hallo Zusammen. Ich bin hier auf diesen thread eigentlich gestoßen weil die letzte 1.73.0 Version der Integration bei mir dazu führte das man von HA aus in meiner eigenen produktiven RaspberryMatic gar nichts mehr schalten konnte. Das ist aber ggf nen anderes Problem

Nun frag dich mich gerade wenn ich mir diesen testscript hier genau anschaue was hier genau die Motivation eigentlich ist? Und ich hab auch im ChangeLog der 1.74.0 Version gesehen das es anscheinend invalide Zeichen geben soll die nicht in descriptions (DPInfo) landen dürfen weil sonst das übergebene JSON invalide wird? Gibt es einen Grund wieso ihr die Zeichen nicht einfach escapen könnt bevor ihr sie übergebt? Oder warum macht ihr nicht einfach mit Urlencode daraus ein URL kodierten String den ihr dann auf HA Seite wieder decodiert? Dann könnte man eigentlich ja jeglichen String rüberschicken weil ja dann " und space und Tab nicht vorkommen können und dann zerlegt die Integration das in seine Bestandteile bzw macht dann ein url decoding, usw. Wäre das nicht ein praktikabler Ansatz? Und wenn URL Encoding nicht reicht könnte man auch ggf. den gesamten String base64 encoden oder so?!?

**Reply by SukramJ** [reply]
Moin Jens,

> das es anscheinend invalide Zeichen geben soll die nicht in descriptions (DPInfo) landen dürfen ...

Ursprung für die Liste  `<	>	'	"	&	$	[	]	{	}	\` ist die Validierungsfehlermeldung in RM, das diese nicht in ein Beschreibungsfeld eingetragen werden dürfen. 

Wobei nur zwei Zeichen (`"	 \`) und TAB für das resultierende JSON ein Problem darstellen. 
Das mit dem codieren/decodieren hört sich gut. Wenn jemand dazu eine entsprechenden PR für [PRG](https://github.com/danielperna84/hahomematic/blob/devel/hahomematic/rega_scripts/get_program_descriptions.fn) und [SV](https://github.com/danielperna84/hahomematic/blob/devel/hahomematic/rega_scripts/get_system_variable_descriptions.fn) einstellt oder das entsprechende Code-Snippet hier einstellt, dann kann ich gerne die HA Gegenseite bauen.

**Reply by jens-maus** [reply]
> > das es anscheinend invalide Zeichen geben soll die nicht in descriptions (DPInfo) landen dürfen ...
> 
> Ursprung für die Liste  `<	>	'	"	&	$	[	]	{	}	\` ist die Validierungsfehlermeldung in RM, das diese nicht in ein Beschreibungsfeld eingetragen werden dürfen. 

Wer definiert denn sowas?!? Also in RaspberryMatic ist das alles kein Problem.

> Wobei nur zwei Zeichen (`"	 \`) und TAB für das resultierende JSON ein Problem darstellen. 
> 
> Das mit dem codieren/decodieren hört sich gut. Wenn jemand dazu eine entsprechenden PR für [PRG](https://github.com/danielperna84/hahomematic/blob/devel/hahomematic/rega_scripts/get_program_descriptions.fn) und [SV](https://github.com/danielperna84/hahomematic/blob/devel/hahomematic/rega_scripts/get_system_variable_descriptions.fn) einstellt oder das entsprechende Code-Snippet hier einstellt, dann kann ich gerne die HA Gegenseite bauen.

Ich schau mir das nachher mal genauer an und versuche ein PR oder code snippet einzustellen. Du müsstest dann halt in eure Integration die decodierung einbauen.

**Reply by SukramJ** [reply]
> Wer definiert denn sowas?!? Also in RaspberryMatic ist das alles kein Problem.

Ich denke Du kennst die Leute besser.

Wenn ich eine Sysvar in RM mit der Beschreibung `Test "bla bla"` anzulegen versuchen kommt die folgende Meldung:
![7A64197C-642B-43E0-99FC-BC983995C881_4_5005_c](https://github.com/user-attachments/assets/6280e07b-1db1-42fd-b9d6-516ed0ff64c9)

Ich bin davon ausgegangen, das dadurch auch sichergestellt wird, das die Beschreibungen "sauber sind, aber man lernt nie aus.

> Ich schau mir das nachher mal genauer an und versuche ein PR oder code snippet einzustellen. 

Danke.

**Reply by jens-maus** [reply]
> > Wer definiert denn sowas?!? Also in RaspberryMatic ist das alles kein Problem.
> 
> Ich denke Du kennst die Leute besser.
> 
> Wenn ich eine Sysvar in RM mit der Beschreibung `Test "bla bla"` anzulegen versuchen kommt die folgende Meldung: ![7A64197C-642B-43E0-99FC-BC983995C881_4_5005_c](https://private-user-images.githubusercontent.com/5650106/395767098-6280e07b-1db1-42fd-b9d6-516ed0ff64c9.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQxNzM2NDQsIm5iZiI6MTczNDE3MzM0NCwicGF0aCI6Ii81NjUwMTA2LzM5NTc2NzA5OC02MjgwZTA3Yi0xZGIxLTQyZmQtYjlkNi01MTZlZDBmZjY0YzkuanBlZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE0VDEwNDkwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdlMTJlOTNhZmU0NTA5OWZjODVmODA5M2Q0OTJlODQ0OWE3NmM5MDQ1YTg5YmU3ZDQwZmFjYjgxMjg1NmJmNjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.H8X22MB4-Ynb_ZchWAOCn-HLcdN437XEycb26QVIP68)

Interessant. Muss ich mir mal anschauen. Aber ich denke das ist nur die Limitation das man das so nicht via WebUI eingeben kann. Das heisst aber nicht das `DPInfo()` und `PrgInfo()` nicht auch diese Zeichen beinhalten kann/könnte.

> Ich bin davon ausgegangen, das dadurch auch sichergestellt wird, das die Beschreibungen "sauber sind, aber man lernt nie aus.
> 
> > Ich schau mir das nachher mal genauer an und versuche ein PR oder code snippet einzustellen.
> 
> Danke.

Siehe https://github.com/danielperna84/hahomematic/pull/1911. D.h. es sollte dann nur noch der entsprechende Python code fehlen (z.b. via `urllib.parse.unquote()` - https://docs.python.org/3/library/urllib.parse.html#urllib.parse.unquote) der den `%XX` codierten String dann wieder entsrpechende decodiert. Zu beachten gilt hier lediglich, das du in dieser Funktion sicherstellen musst das die eingangsdaten nicht `utf-8` sind/waren, sondern die rega bzw. die ccu ja immer mit `iso-8859-1` strings hantiert und du daher in dieser funktion den `encoding` parameter auch entsprechend auf `iso-8859-1` stellen musst bei der decodierung.

**Comment by SukramJ** [comment]
Bitte mal mit der 1.75.0b0 testen.

**Comment by Anybody4u** [comment]
Hallo, vielen Dank für die Infos und das Script, das hat sehr geholfen. Gleichwohl nach Entfernung der ganannten Zeichen <	> ' "	& $	[ ] { } \   hatte ich immer noch die Fehlermeldung. Ich habe dann auch noch die Umlaute Ä Ö Ü und den Unterstrich _ entfernt um den Fehler zu beheben. Leider habe ich Umlaute und Unterstrich gemeinsam entfernt, so dass ich jetzt nicht weiß, ob es an den Umlauten, dem Unterstrich oder Beidem gelegen hat.

**Reply by SukramJ** [reply]
Benutz du die  1.75.0 ?

**Reply by Anybody4u** [reply]
Ich bentutze Homematic Integration 1.75.0
RaspberryMatic 3.79.6.20241122
Home Assistant Core 2024.12.3

**Reply by SukramJ** [reply]
Dann sollte es mit den Sysvars keine Probleme mehr geben.
